File: keepalive.js
#!/usr/bin/env node

/**
 * Ultra-lightweight 24/7 Keep-Alive Script
 * Pings your web server with zero load (using HEAD/GET requests)
 * to prevent cloud hosting (Render, Koyeb, Glitch, etc.) from sleeping.
 *
 * Usage:
 *   node keepalive.js [URL] [INTERVAL_MINUTES]
 * Example:
 *   node keepalive.js https://imagegenprompt.onrender.com/ 10
 */

const https = require('https');
const http = require('http');

const TARGET_URL = process.argv[2] || process.env.TARGET_URL || 'https://imagegenprompt.onrender.com/';
const INTERVAL_MIN = parseFloat(process.argv[3] || process.env.INTERVAL_MIN || '10');
const INTERVAL_MS = Math.max(1, INTERVAL_MIN) * 60 * 1000;

function pingServer() {
  const parsed = new URL(TARGET_URL);
  const client = parsed.protocol === 'https:' ? https : http;

  const req = client.request(
    TARGET_URL,
    {
      method: 'HEAD',
      headers: {
        'User-Agent': 'KeepAliveBot/1.0 (Zero-Load HealthCheck)',
      },
      timeout: 15000,
    },
    (res) => {
      const timestamp = new Date().toISOString();
      console.log(`[${timestamp}] Pinged ${TARGET_URL} - Status: ${res.statusCode}`);
      // Consume response data to free up memory
      res.resume();
    }
  );

  req.on('timeout', () => {
    console.warn(`[${new Date().toISOString()}] Warning: Request timed out for ${TARGET_URL}`);
    req.destroy();
  });

  req.on('error', (err) => {
    console.error(`[${new Date().toISOString()}] Ping error:`, err.message);
  });

  req.end();
}

console.log('='.repeat(60));
console.log('  Zero-Load 24/7 Keep-Alive Bot Started');
console.log(`  Target URL : ${TARGET_URL}`);
console.log(`  Interval   : Every ${INTERVAL_MIN} minutes`);
console.log('='.repeat(60));

// Initial immediate ping
pingServer();

// Continuous infinite loop
setInterval(pingServer, INTERVAL_MS);

File: keepalive.py
#!/usr/bin/env python3
"""
Ultra-lightweight 24/7 Keep-Alive Script (Python)
Zero dependencies - uses built-in urllib.
Pings the target URL using lightweight HEAD requests to keep servers alive infinitely.

Usage:
    python3 keepalive.py [URL] [INTERVAL_MINUTES]
"""

import sys
import time
import urllib.request
import urllib.error
from datetime import datetime

TARGET_URL = sys.argv[1] if len(sys.argv) > 1 else "https://imagegenprompt.onrender.com/"
INTERVAL_MINUTES = float(sys.argv[2]) if len(sys.argv) > 2 else 10.0
INTERVAL_SECONDS = max(60, int(INTERVAL_MINUTES * 60))

class HeadRequest(urllib.request.Request):
    def get_method(self):
        return "HEAD"

def ping():
    timestamp = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
    req = HeadRequest(
        TARGET_URL,
        headers={"User-Agent": "KeepAliveBot/1.0 (Zero-Load HealthCheck)"}
    )
    try:
        with urllib.request.urlopen(req, timeout=15) as response:
            print(f"[{timestamp}] Pinged {TARGET_URL} - Status: {response.status}")
    except urllib.error.HTTPError as e:
        print(f"[{timestamp}] Pinged {TARGET_URL} - HTTP Status: {e.code}")
    except Exception as e:
        print(f"[{timestamp}] Ping error: {e}")

if __name__ == "__main__":
    print("=" * 60)
    print("  Zero-Load 24/7 Python Keep-Alive Bot Started")
    print(f"  Target URL : {TARGET_URL}")
    print(f"  Interval   : Every {INTERVAL_MINUTES} minutes")
    print("=" * 60)
    
    while True:
        ping()
        time.sleep(INTERVAL_SECONDS)
